File: 146. LRU Cache.py
# Problem Link: https://leetcode.com/problems/lru-cache/
class Node:
    def __init__(self, key, val, prev=None, nxt=None):
        self.key = key
        self.val = val
        self.prev = prev
        self.nxt = nxt


class LRUCache:

    def __init__(self, capacity: int):
        self.cache = {}
        self.cap = capacity
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.nxt = self.tail
        self.tail.prev = self.head

    def add(self, n):
        prev = self.tail.prev
        nt = self.tail
        prev.nxt = n
        nt.prev = n
        n.nxt = nt
        n.prev = prev

    def remove(self, n):
        prev, nxt = n.prev, n.nxt
        prev.nxt = nxt
        nxt.prev = prev

    def get(self, key: int) -> int:
        if key in self.cache:
            self.remove(self.cache[key])
            self.add(self.cache[key])
            return self.cache[key].val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.remove(self.cache[key])
        self.cache[key] = Node(key, value)
        self.add(self.cache[key])

        if len(self.cache) > self.cap:
            lru = self.head.nxt
            self.remove(lru)
            del self.cache[lru.key]
            del lru

File: C#/146. LRU Cache.cs
public class Node {
    public int key, val;
    public Node prev, next;
    
    public Node(int key, int val, Node prev=null, Node next=null){
        this.key = key;
        this.val = val;
        this.prev = prev;
        this.next = next;
    }
}

public class LRUCache {
    public Dictionary<int, Node> cache = new Dictionary<int, Node>();
    int cap;
    Node head, tail;
    
    public LRUCache(int capacity) {
        cap = capacity;
        head = new Node(0, 0);
        tail = new Node(0, 0);
        head.next = tail;
        tail.prev = head;
    }
    
    public void Add(Node n){
        var prv = tail.prev;
        prv.next = n;
        n.next = tail;
        tail.prev = n;
        n.prev = prv;
    }
    
    public void Remove(Node n){
        var prv = n.prev;
        var nxt = n.next;
        prv.next = nxt;
        nxt.prev = prv;
    }
    
    public int Get(int key) {
        if (cache.ContainsKey(key)){
            Remove(cache[key]);
            Add(cache[key]);
            return cache[key].val;
        }
        
        return -1;
    }
    
    public void Put(int key, int value) {
        if (cache.ContainsKey(key)){
            Remove(cache[key]);
        }
        cache[key] = new Node(key, value);
        Add(cache[key]);
        
        if (cache.Count > cap){
            var lru = head.next;
            Remove(lru);
            cache.Remove(lru.key);
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.Get(key);
 * obj.Put(key,value);
 */